envoy/network/connection_handler.h

```diff
@@ -136,6 +139,12 @@ class ConnectionHandler {
      */
     virtual void shutdownListener() PURE;
 
+    /**
+     * Inject optional custom handling of UDP packets during hot restart.
+     * It is up to the implementation to capture and call the provided function.
+     */
+    virtual void onHotRestarting(HotRestartPacketForwardingFunction) {}
```

Comment:
Maybe, if we're doing forwarding only parent->child (and child not actually listening directly until drain is complete), as discussed in the thread. I'll have to look into it a bit, I think that it might not be great to complicate the common interface for the rare case.

Reply:
I guess that could be okay if shutdownListener takes some sort of abstract shared object parameter, then individual listeners could try to pull out `dynamic_pointer_cast<SomethingThisListenerCaresAbout>`.
Like `shutdownListener(std::shared_ptr<AdditionalShutdownParameters>)`
and `class HotRestartShutdownParameters : public AdditionalShutdownParameters, public HotRestartPacketForwardingProvider`

Putting *specifically* a hot restart parameter in shutdownListener, that only Quic listeners actually care about, seems like it would be more confusing even than this tangled dynamic_cast object (and much more horrible for the risk of future custom arguments that some other listener might want - the dynamic object at least makes it flexible for delivering whatever sort of arguments any listener might be interested in).

Is that better than a new virtual function that for most listeners is a no-op? I can see how maybe because it saves on piping a new call to the virtual function through all the various layers of controllers. But changing shutdownListener is still going to require a change to all those same layers, *this time*, at least.

Reply:
Where is onHotRestarting() going to be called? Can the call site distinguish what listener it is calling into?

Reply:
I was planning to call it from somewhere around the [hot restarting functions](https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/server/hot_restarting_parent.cc?L102-132) where the socket ids are collected. The idea is the call site wouldn't need to distinguish which listeners need it, because all the *other* listeners have no-op implementations.

Reply:
Made a try with a parameter to shutdownListeners, to see what you think.

Reply:
Thanks for the code pointer! It seems that it's seperate RPC calls to PassListenSocket and DrainListeners: https://sourcegraph.com/github.com/envoyproxy/envoy/-/blob/source/server/hot_restarting_parent.cc?L52. Though it makes more sense to tell the listener whether/how to forward packets while telling it to drain, but it's not very good to collect socket ids twice once. So I don't strong opinion on this. 

But we do want to make sure the sequence is right so that the parent instance knows how to forward the packets once it starts draining. Can we enforce the sequence on child side?

Reply:
I'm still working on this aspect in the next PR. Since I've changed it to passing in shutdownListeners like you asked, there's now no choice about where to provoke the change, it has to be drainListeners. It's also pretty awkward on the child side now, because with this "only parent to child" forwarding (rather than the original plan of two-way), hitting the right moment for the child to start listening is tricky - every other socket starts listening immediately, but this one will now have to hold off on listening until a moment before the parent drain completes. (On the up side, the location of this awkwardness is the same as the location where we'll need to do something awkward for the eBPF version, so at least the effort is dual-purpose.)

This part of the conversation should probably move to slack or the relevant [issue](https://github.com/envoyproxy/envoy/issues/19454) though, since it's not important to this PR. :)

Reply:
Having the parent say when it's done would be a dangerous choice, in that if the parent exited prematurely the child will *never* start listening, plus there's no existing channel for this form of communication (though it could piggyback on the packet forwarding channel which is also new).

My current thought is that the child starting listening immediately before or after it tells the parent to terminate would make sense, and be much simpler. (Doesn't matter if it's before or after because the socket *exists*, so it can still buffer up some content after the other instance closes, and if it picks up a few packets that were intended for connections associated with the parent it will reject them, which is okay because the parent is about to terminate and reject all packets anyway.)